src/get-installation-url.ts

```diff
@@ -0,0 +1,37 @@
+import type { App } from "./index.js";
+import type { GetInstallationUrlOptions } from "./types.js";
+
+export function getInstallationUrlFactory(app: App) {
+  let installationUrlBasePromise: Promise<string> | undefined;
+
+  return async function getInstallationUrl(options: GetInstallationUrlOptions) {
+    if (!installationUrlBasePromise) {
+      installationUrlBasePromise = getInstallationUrlBase(app);
+    }
+
+    const installationUrlBase = await installationUrlBasePromise;
+    const installationUrl = new URL(installationUrlBase);
+
+    if (options.target_id !== undefined) {
+      installationUrl.pathname += "/permissions";
+      installationUrl.searchParams.append(
+        "target_id",
+        options.target_id.toFixed(),
+      );
+    }
+
+    if (options.state !== undefined) {
+      installationUrl.searchParams.append("state", options.state);
+    }
+
+    return installationUrl.href;
+  };
+}
+
+async function getInstallationUrlBase(app: App) {
+  const { data: appInfo } = await app.octokit.request("GET /app");
+  if (!appInfo) {
+    throw new Error("[@octokit/app] unable to fetch metadata for app");
+  }
+  return `${appInfo.html_url}/installations/new`;
```

Comment:
I don't think there is a case where `GET /app` succeeds but the response body is falsy. Either the request succeeds or it fails. I don't think we need extra error handling here, it will just propagate and should be clear enough by itself.

```suggestion
  const { data  } = await app.octokit.request("GET /app");
  return `${data.html_url}/installations/new`;
```

Reply:
Yeah this extra check is almost exclusively to satisfy the types. The type of `appInfo` is `{ html_url: string; ... } | null`. I didn't venture to look at the source to see if `null` is actually possible, but I opted to just make typescript happy

Reply:
okay I'll look into this. Most likely it's a bug in the OpenAPI definition from which the types are derived automatically.